File: rob599_basic/rob599_basic/twice.py
#!/usr/bin/env python3


# Example of a simple transformer node in ROS 2
#
# twice.py
#
# Bill Smart
#
# This node subscribes to a topic with Int64s.  It multiplies the numbers by a constant,
# and then republishes them on another topic.


# Every Python node in ROS2 should include these lines.  rclpy is the basic Python
# ROS2 stuff, and Node is the class we're going to use to set up the node.
import rclpy
from rclpy.node import Node

# We're going to publish an Int64.  This is the same as ROS.
from std_msgs.msg import Int64


class Twice(Node):
	'''
	This node subscribes to a topic publishing Int64s, doubles each number if receives, and
	# repubishes the result on an outbound topic.
	'''
	def __init__(self):
		# Initialize the parent class, giving it a name.  The idiom is to use the
		# super() class.
		super().__init__('twice')

		# Create the publisher first, to make sure it's available before we start the subscriber.
		self.pub = self.create_publisher(Int64, 'doubled', 10)

		# Create the subscriber.
		self.sub = self.create_subscription(Int64, 'number', self.callback, 10)


	# This callback will be called whenever we receive a new message on the topic.
	def callback(self, msg):
		# Create a new message.
		new_msg = Int64()

		# Multiply the data element of the message.
		new_msg.data = msg.data * 2

		# Make sure we record what's going on
		self.get_logger().info(f'Got {msg.data} and transformed it to {new_msg.data}')

		# Republish the message.  Reusing the original message is slightly more efficient
		# than creating a new one.
		self.pub.publish(new_msg)


# This is a entry point.	
def main(args=None):
	# Initialize rclpy.  We should do this every time.
	rclpy.init(args=args)

	# Make a node class.  The idiom in ROS2 is to encapsulte everything in a class
	# that derives from Node.
	twice = Twice()

	# The spin() call gives control over to ROS2, and it now takes a Node-derived
	# class as a parameter.
	rclpy.spin(twice)

	# Make sure we shutdown everything cleanly.  This should happen, even if we don't
	# include this line, but you should do it anyway.
	rclpy.shutdown()


# If we run the node as a script, then we're going to start here.
if __name__ == '__main__':
	# The idiom in ROS2 is to set up a main() function and to call it from the entry
	# point of the script.
	twice()

File: rob599_basic_cpp/src/twice.cpp
// twice.cpp
//
// Bill Smart
//
// An example of a simple C++ transformer node in ROS 2.  This node subscribes to a topic
// with Int64 messages, doubles the value, and republishes.

// Include the basic ROS functionality and the message definition.
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/int64.hpp>

// We need this in order to set up the subscriber.  This is the C++ way of
// letting us specify callback parameters.
using std::placeholders::_1;


// As usual, we create a class that inherits from Node.
class Twice : public rclcpp::Node {
public:
	Twice() :Node("twice") {
		// Set up the publisher before the subscriber.  The type goes in the template
		// specification, and the queue size is 10.
		publisher_ = this->create_publisher<std_msgs::msg::Int64>("doubled", 10);

		// Set up the subscriber.  The type goes in the template specification.  The callback
		// argument uses std::bind to combine a pointer to the callback and a placeholder for
		//the single argument.
		subscriber_ = this->create_subscription<std_msgs::msg::Int64>("number", 10, std::bind(&Twice::callback, this, _1));
	}

private:
	// Variables for the publisher and subscriber.
	rclcpp::Subscription<std_msgs::msg::Int64>::SharedPtr subscriber_;
	rclcpp::Publisher<std_msgs::msg::Int64>::SharedPtr publisher_;

	// The callback for the subscriber.  The message is passed via a shared pointer.  This means
	// you have to use the -> accessor, rather than the . (dot) accessor.
	void callback(const std_msgs::msg::Int64::SharedPtr msg) const {
		// The type is clear, so we can use the auto keyword.
		auto doubled = std_msgs::msg::Int64();
		doubled.data = msg->data * 2;

		// Publish the message.
		publisher_->publish(doubled);

		// Write to the log.
		RCLCPP_INFO(this->get_logger(), "Got %li and transformed it to %li", msg->data, doubled.data);
	}
};


int main(int argc, char **argv) {
	// Initialize ROS.
	rclcpp::init(argc, argv);

	// Create a node.
	auto doubler = std::make_shared<Twice>();

	// Give control to the ROS event handler.
	rclcpp::spin(doubler);

	// Once the event handler is done, shut things down nicely.
	rclcpp::shutdown();

	return 0;
}